proc.c

```diff
@@ -628,3 +631,59 @@ procdump(void)
       cprintf("\n");
   }
 }
+
+
+int kthread_create(void (*start_func)(), void* stack){
+    //TODO
+    return 0;
+}
+
+int kthread_id(){
+    return my_thread()->tid;
+}
+
+void kthread_exit(){
+    //struct proc *curproc = myproc();
+    struct thread * curr_thread = my_thread();
+    boolean lastthread = TRUE;
+    struct proc* proc = myproc();
+    acquire(&proc->proclock);
+    struct thread * t;
+    for (t = proc->threads; t < &proc->threads[NTHREADS]; t++){
+        if(t->t_state == T_RUNNABLE | t->t_state==T_RUNNING |t->t_state == T_SLEEPING) {
+            lastthread = FALSE;
+            break;
+        }
+
+    }
+    if(lastthread){
+        exit();
+    }else {
+        curr_thread->t_state = T_TERMINATED;
+        sched();
```

Comment:
before entering sched you must have both ptable and proc lock acquired. and it must be in a specific order (ptable first)

Reply:
you must handle trap.c and add a check there for each thread, to see if it's state is not T_terminate.